### ISSUE TITLE
Disconnect between evaluation and processing

### ISSUE THREAD

**Issue by neubig** [neulab/xnmt#260]
There is a problem with evaluation of likelihood and BLEU at dev time when using output tokenized by, for example, sentence piece. Tokenized output files are necessary for evaluation of likelihood, but untokenized output files are necessary for evaluation of BLEU. Right now we can't specify both.

I would like to propose an overhaul about how we think about evaluation. Each type of evaluation (e.g. evaluation of perplexity, evaluation of MT metrics) each gets its own class. These evaluation tasks can take in different reference files. We don't include the detokenizer in the inference engine, and instead it lives in the evaluation task. These evaluation tasks could be shared between dev time and test time by using the component sharing mechanism if necessary.

What do you think @msperber ?

**Comment by msperber** [comment]
Yes, it seems the evaluation approach could use some restructuring. I think that (after correctness obviously) one goal should be to keep the config files as simple as possible. I attached an example config, is that what you were thinking of? There are a couple of problems with this particular version though:
- it's difficult to avoid copying the evaluators between dev and test because they need different reference files
- some evaluators might need several post-processing steps (such as join-char and detokenize), how should we go about supporting that? Maybe reusing some of the infrastructure from preproc.py?
- to support the schedule_metric, instead of one 'evaluators' attribute we might to need to: 'schedule_evaluator' and 'auxiliary_evaluators'

```exp:
  experiment: ...
  train: !SimpleTrainingRegimen
    evaluators:
    - !BleuEvaluator
      post_process: join-char
      ref: dev.trg.words
    corpus_parser: !BilingualCorpusParser
      src_reader: !PlainTextReader {}
      trg_reader: !PlainTextReader {}
      training_corpus: !BilingualTrainingCorpus
        train_src: train.src
        train_trg: train.trg.char
        dev_src: dev.src
        dev_trg: dev.trg.char
    model: ...
  inference: !SimpleInference
    src_file: examples/output/test.norm.ja
  test-evaluators:
  - !BleuEvaluator
    post_process: join-char
    ref: test.trg.words
```

**Comment by neubig** [comment]
OK, I've thought about this a bit, and here is a proposed model format, which I'll explain a bit below the fold.

    experiment: ...
      model:
        ... # model parameters
        inference: !SimpleInference
          beam: 5
        src_reader: !PlainTextReader {}
        trg_reader: !PlainTextReader {}
      train: !SimpleTrainingRegimen
        ... # training parameters
        src_file: train.src
        trg_file: train.trg.chars
        dev_tasks:
          - !AccuracyEvalTask
            eval_metrics: bleu wer
            post_process: join-char
            src_file: dev.src
            trg_file: dev.trg.words
          - !LossEvalTask
            src_file: dev.src
            trg_file: dev.trg.chars
      test:
        - !AccuracyEvalTask
          eval_metrics: bleu wer
          post_process: join-char
          src_file: test.src
          trg_file: test.trg.words
        - !DecodingTask
          post_process: join-char
          src_file: blind.src

Here are some details and assumptions:
* We will always have a "source" and a "target". I think this doesn't hurt generality at all, because our "source" could actually be multiple things, and our "target" could also be multiple things if we want.
* The model knows how to read the source and target files, and only allows for one format for this input and output.
* An EvalTask returns one or more EvalScores.
* There is a "dev_combinator:" option in train that defaults to "x[0]", which means that the scheduling is done based on the first metric calculated by the first dev_task. Optionally, you could specify something like "x[0]-x[1]" which would mean that scheduling would be done based on BLEU-WER (we'd need to think about the cleanest way to deal with things like WER where lower is better).
* "train:" has an optional "model:" parameter that defaults to "experiment.model"
* EvalTasks have an optional "model:" parameter that defaults to "self.parent().model" (not sure how we get the parent in the hierarchy?)
* AccuracyEvalTask and DecodingTask have a optional "inference:" parameter that defaults to "self.model.inference"

Here is an example of what this would look like in a multi-task setting where we train an MT system and POS tagger with a shared encoder.

    experiment: ...
      train: !SameBatchMultiTaskTrainingRegimen
        tasks:
          - !SimpleTrainingRegimen
            model: &mt_model
              ... # model parameters
              encoder: &input_encoder
                ...
              src_reader: !PlainTextReader &input_reader {}
              trg_reader: !PlainTextReader {}
            src_file: mt-train.src
            trg_file: mt-train.trg
          - !SimpleTrainingRegimen
            model: &pos_model
              ... # model parameters
              encoder: *input_encoder
                ...
              src_reader: *input_reader
              trg_reader: !PlainTextReader {}
            src_file: pos-train.src
            trg_file: pos-train.trg
        dev_tasks:
          - !AccuracyEvalTask
            model: *mt_model
            eval_metrics: bleu
            src_file: mt-dev.src
            trg_file: mt-dev.trg
          - !AccuracyEvalTask
            model: *pos_model
            eval_metrics: wer
            src_file: pos-dev.src
            trg_file: pos-dev.trg
        eval_combinator: "x[0]-x[1]"
      test:
        - !AccuracyEvalTask
          model: *mt_model
          eval_metrics: bleu
          src_file: mt-test.src
          trg_file: mt-test.trg
        - !AccuracyEvalTask
          model: *pos_model
          eval_metrics: wer
          src_file: pos-test.src
          trg_file: pos-test.trg

What do you think? This is a major change, but I think it would make things conceptually simpler. If you think this is worth doing, I might make a stab at doing this, starting with first restructuring the config files to make sure I didn't miss anything conceptually, then by actually doing the implementation.

**Comment by neubig** [comment]
Sorry, I realize that I didn't answer your first two questions directly.

* Multiple post-processors: how about just listing them in sequence?
* Copying evaluators between dev/test: I think this actually may be a feature, not a bug. You'll notice that this is convenient if we want to do stuff like decoding a blind test set, or perhaps running a different variety of evaluation at test time, etc.

**Comment by msperber** [comment]
I think that looks good and more intuitive than the current design. One thing to consider would be whether we would need to extend the YAML (de)serializer. Basically, I don't see any problem at the moment, but it would probably be good to keep in mind while restructuring the config files. I can help with the serialization if necessary. For reference, the basic process is this:
- Create stubs (uninitialized objects) from the .yaml file
- share parameters between stubs (top down)
- initialize stubs bottom up, while resolving the DependentInitParams and reference-shared components

By the way, in the examples, you are only doing config-sharing when it should be reference-sharing (add a ```_xnmt_id``` to the base component). This has been something that bothered me while implementing the component sharing: I would prefer getting rid of the "config-sharing" thing and only support reference-sharing even without having to set an ```_xnmt_id```. I haven't found a good way for doing that because the YAML parser doesn't tell us if things are referenced or specified explicitly, but it might be worth revisiting this if necessary.

**Comment by msperber** [comment]
For the parent() feature: it might be possible to do that with a DependentInitParam, if not that would need to be implemented in the deserializer.

**Comment by neubig** [comment]
+1 on getting rid of "config sharing" if possible.

And thanks, let me get as far as I can without changing the serializer, then ask for your help. Even without the serializer, I can do anything I mentioned here with the current serializer by adding component sharing pointers as appropriate.

**Comment by neubig** [comment]
Fixed by #267
